napalm_logs/base.py

```diff
@@ -130,6 +132,77 @@ def _load_config(self, path):
             raise IOError(msg)
         return config
 
+    @staticmethod
+    def _raise_config_exception(error_string):
+        log.error(error_string, exc_info=True)
+        raise MissConfigurationException(error_string)
+
+    def _verify_config_key(self, key, value, valid, config, dev_os, key_path):
+        key_path.append(key)
+        if not config.get(key):
+            self._raise_config_exception('Unable to find key "{}" for {}'.format(':'.join(key_path), dev_os))
+        if isinstance(value, type):
+            if not isinstance(config[key], value):
+                self._raise_config_exception('Key "{}" for {} should be {}'.format(':'.join(key_path), dev_os, value))
+        elif isinstance(value, dict):
+            if not isinstance(config[key], dict):
+                self._raise_config_exception('Key "{}" for {} should be of type <dict>'.format(':'.join(key_path), dev_os))
+            self._verify_config_dict(value, config[key], dev_os, key_path)
+            # As we have already checked that the config below this point is correct, we know that "line" and "values"
+            # exists in the config if they are present in the valid config
+            if 'line' in value.keys() and 'values' in value.keys():
+                from_line = re.findall('\{(\w+)\}', config[key]['line'])
+                if set(from_line) != set(config[key]['values']):
+                    self._raise_config_exception('The "values" do not match variables in "line" for {} in {}'.format(':'.join(key_path), dev_os))
+        key_path.remove(key)
+
+    def _verify_config_dict(self, valid, config, dev_os, key_path=None):
+        if not key_path:
+            key_path = []
+        for key, value in valid.items():
+            # If the key is '*' then we should check all keys in the config to make sure they match the allowed values
+            if key == '*':
+                for config_key in config.keys():
+                    self._verify_config_key(config_key, value, valid, config, dev_os, key_path)
+            else:
+                self._verify_config_key(key, value, valid, config, dev_os, key_path)
+
+    def _verify_config(self):
+        '''
+        Verify that the config is correct
+        '''
+        if not self.config_dict:
+            self._raise_config_exception('No config found')
+
+        valid_opts = {
```

Comment:
I have made the variable global as it makes sense for it to be global, but I still have to pass it to `self._verify_config_dict` the first time it is called. It is a recursive function and each time the valid config for the specific section is passed.